src/journal/journald-server.c

```diff
@@ -1378,7 +1378,7 @@ int server_process_datagram(sd_event_source *es, int fd, uint32_t revents, void
 
         if (fd == s->syslog_fd) {
                 if (n > 0 && n_fds == 0)
-                        server_process_syslog_message(s, strstrip(s->buffer), ucred, tv, label, label_len);
+                        server_process_syslog_message(s, s->buffer + strspn(s->buffer, WHITESPACE), ucred, tv, label, label_len);
```

Comment:
I went ahead and implemented @poettering's suggestion. However, I don't know how to test it. glibc's syslog implementation seems to drop everything after encountering first NULL bytes. Hence journald never receives a message that contains embedded NULL bytes.